include/targets/cuda/reduce_helper.h

```diff
@@ -167,76 +151,137 @@ namespace quda
     }
   };
 
+  /**
+     @brief Write the reduced result out.  Three different destinations are supported:
+       * If arg.result_d is non-null we atomically write out the
+         result (to host-mapped memory)
+       * Else if arg.get_output_async_buffer returns non-zero we
+         non-atomically write out the result to device memory
+       * Else we atomically write out the result to device memory
+     @param[in,out] arg Kernel argument
+     @param[in] sum Reduced value to be written out
+     @param[in] idx Memory index we are writing to
+   */
+  template <typename Arg, typename T>
+  __device__ inline void write_result(Arg &arg, const T &sum, const int idx)
+  {
+    using atomic_t = typename atomic_type<T>::type;
+    constexpr size_t n = sizeof(T) / sizeof(atomic_t);
+
+    if (arg.result_d) { // write to host mapped memory
+#ifdef _NVHPC_CUDA // WAR for nvc++
+      constexpr bool coalesced_write = false;
+#else
+      constexpr bool coalesced_write = true;
+#endif
+      if constexpr (coalesced_write) {
+        static_assert(n <= device::warp_size(), "reduction array is greater than warp size");
+        auto mask = __ballot_sync(0xffffffff, target::thread_idx().x < n);
+        if (target::thread_idx().x < n && target::thread_idx().y == 0) {
+          atomic_t sum_tmp[n];
+          memcpy(sum_tmp, &sum, sizeof(sum));
+
+          atomic_t s = sum_tmp[0];
+#pragma unroll
+          for (int i = 1; i < n; i++) {
```

Comment:
I believe there's a corner case where this could break if `blockDim.x < n <= device::warp_size()` --- is this a situation worth worrying about?

Reply:
7eba3b448a0f8cc73db93469991e3d6ae0b911d0